pkg/querier/store.go

```diff
@@ -92,9 +113,99 @@ outer:
 			}
 			iterators = append(iterators, iterator)
 		}
+
 		result = append(result, iter.NewNonOverlappingIterator(iterators, labels))
 	}
-	return result, nil
+
+	return iter.NewHeapIterator(result, req.Direction), nil
+}
+
+func loadFirstChunks(ctx context.Context, chks map[model.Fingerprint][][]chunkenc.LazyChunk) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "loadFirstChunks")
+	defer sp.Finish()
+
+	// If chunks span buckets, then we'll have different fetchers for each bucket.
+	chksByFetcher := map[*chunk.Fetcher][]*chunkenc.LazyChunk{}
+	for _, lchks := range chks {
+		for _, lchk := range lchks {
+			chksByFetcher[lchk[0].Fetcher] = append(chksByFetcher[lchk[0].Fetcher], &lchk[0])
+		}
+	}
+
+	sp.LogFields(otlog.Int("fetchers", len(chksByFetcher)))
+
+	errChan := make(chan error)
+	for fetcher, chunks := range chksByFetcher {
+		go func(fetcher *chunk.Fetcher, chunks []*chunkenc.LazyChunk) {
```

Comment:
it seems that you expect the first chunk to be available right after this method end as you filter using Metric property from chunk.

However you don't wait for all goroutines to terminate before returning, not sure if this is intended.

Reply:
Ho ! you are using the errChan to wait, still you have some code after that sets chunk value that won't be waited for execution.

One way to fix this is to change the code below like this:

```golang
chks, err := fetcher.FetchChunks(ctx, chks, keys)
			
			if err != nil {
                             errChan <- err
				return
			}

 			for i, chk := range chks {
				chunks[i].Chunk = chk
			}
                         errChan <- nil
```

Reply:
Nice catch!